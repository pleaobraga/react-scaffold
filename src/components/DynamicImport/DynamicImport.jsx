import React from 'react'
import PropTypes from 'prop-types'

class DynamicImport extends React.Component {
  state = {
    component: null
  }

  componentDidMount() {
    const { loadComponent, ErrorComponent } = this.props

    loadComponent()
      .then(comp => {
        comp.default
          ? this.setState({ Component: comp.default })
          : this.setState({ Component: ErrorComponent })
      })
      .catch(() => this.setState({ Component: ErrorComponent }))
  }

  render() {
    const { loadingComponent } = this.props
    const { Component } = this.state

    return Component ? <Component /> : loadingComponent
  }
}

DynamicImport.defaultProps = {
  ErrorComponent: <div id="error">Error</div>,
  loadingComponent: <div id="loading">Loading...</div>
}

DynamicImport.propTypes = {
  loadComponent: PropTypes.func.isRequired,
  loadingComponent: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
  ErrorComponent: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
}

export default DynamicImport
