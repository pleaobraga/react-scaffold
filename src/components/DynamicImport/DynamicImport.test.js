import 'babel-polyfill'
import React from 'react'
import { mount } from 'enzyme'
import DynamicImport from './DynamicImport'

const a = () => <div id="content">test</div>

const request = data => {
  return new Promise((resolve, reject) => {
    resolve({ default: a })
    // process.nextTick(() =>
    //   data
    //     ? resolve({ default: a })
    //     : reject({
    //       error: 'Error'
    //     })
    // )
  })
}

describe.only('DynamicImport', () => {
  it('render proprely', () => {
    const dynamicImport = mount(
      <DynamicImport loadComponent={() => request('test')} />
    )

    expect(dynamicImport).toMatchSnapshot()
    expect(dynamicImport.find('#loading').length).toBe(1)
  })

  it('render proprely', async () => {
    const dynamicImport = await mount(
      <DynamicImport loadComponent={() => request('test')} />
    )

    dynamicImport.update()

    expect(dynamicImport).toMatchSnapshot()
    expect(dynamicImport.contains(<div id="content">test</div>)).toEqual(true)
  })
})
