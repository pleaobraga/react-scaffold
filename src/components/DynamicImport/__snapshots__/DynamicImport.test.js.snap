// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicImport render proprely 1`] = `
<DynamicImport
  ErrorComponent={
    <div
      id="error"
    >
      Error
    </div>
  }
  loadComponent={[Function]}
  loadingComponent={
    <div
      id="loading"
    >
      Loading...
    </div>
  }
>
  <div
    id="loading"
  >
    Loading...
  </div>
</DynamicImport>
`;

exports[`DynamicImport render proprely 2`] = `
<DynamicImport
  ErrorComponent={
    <div
      id="error"
    >
      Error
    </div>
  }
  loadComponent={[Function]}
  loadingComponent={
    <div
      id="loading"
    >
      Loading...
    </div>
  }
>
  <a>
    <div
      id="content"
    >
      test
    </div>
  </a>
</DynamicImport>
`;
